pkg/helm/actions/install_chart.go

```diff
@@ -109,3 +109,93 @@ func InstallChart(ns, name, url string, vals map[string]interface{}, conf *actio
 	}()
 	return release, nil
 }
+
+func InstallChartAsync(ns, name, url string, vals map[string]interface{}, conf *action.Configuration, client dynamic.Interface, coreClient corev1client.CoreV1Interface, fileCleanUp bool, indexEntry string) (*Secret, error) {
+	var err error
+	var chartInfo *ChartInfo
+	var cp, chartLocation string
+	cmd := action.NewInstall(conf)
+	// tlsFiles contain references of files to be removed once the chart
+	// operation depending on those files is finished.
+	tlsFiles := []*os.File{}
+	if indexEntry == "" {
+		chartInfo, err = getChartInfoFromChartUrl(url, ns, client, coreClient)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		chartInfo = getChartInfoFromIndexEntry(indexEntry, ns, url)
+	}
+
+	connectionConfig, isClusterScoped, err := getRepositoryConnectionConfig(chartInfo.RepositoryName, ns, client)
+	if err != nil {
+		return nil, err
+	}
+
+	if isClusterScoped {
+		clusterConnectionConfig := connectionConfig.(v1beta1.ConnectionConfig)
+		tlsFiles, err = setUpAuthentication(&cmd.ChartPathOptions, &clusterConnectionConfig, coreClient)
+		if err != nil {
+			return nil, fmt.Errorf("error setting up authentication: %v", err)
+		}
+	} else {
+		namespaceConnectionConfig := connectionConfig.(v1beta1.ConnectionConfigNamespaceScoped)
+		tlsFiles, err = setUpAuthenticationProject(&cmd.ChartPathOptions, &namespaceConnectionConfig, coreClient, ns)
+		if err != nil {
+			return nil, fmt.Errorf("error setting up authentication: %v", err)
+		}
+	}
+	cmd.ReleaseName = name
+	if len(tlsFiles) == 0 {
+		chartLocation = url
+	} else {
+		chartLocation = chartInfo.Name
+	}
+	cmd.ChartPathOptions.Version = chartInfo.Version
+	cp, err = cmd.ChartPathOptions.LocateChart(chartLocation, settings)
+	if err != nil {
+		return nil, fmt.Errorf("error locating chart: %v", err)
+	}
+	ch, err := loader.Load(cp)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add chart URL as an annotation before installation
+	if ch.Metadata == nil {
+		ch.Metadata = new(chart.Metadata)
+	}
+	if ch.Metadata.Annotations == nil {
+		ch.Metadata.Annotations = make(map[string]string)
+	}
+	ch.Metadata.Annotations["chart_url"] = url
+
+	cmd.Namespace = ns
+	installComplete := make(chan bool)
+	go func() {
+		_, err := cmd.Run(ch, vals)
+		if err != nil {
+			createSecret(ns, name, 1, coreClient, err)
+		}
```

Comment:
Is it correct to run the `cmd` before creating the `Secret`?

I would expect that we create the Secret on the initial thread and wait for it. So that we don't need to wait for it (getSecret below). And then just running the cmd in a go routine. Wdyt?

Reply:
In the utility.go we are actually wtaching for the secret created by helm to keep a track of helm release status. If an error occurs and we keep watching for the secret we would either get a timeout error or response would be the one as mentioned in get secret as release not found. In order to return the actual result of the error i  am creating this secret , so that the result of cmd.Run can be returned to ui and we are not blocked on the watcher

Reply:
The actual logic is to watch for the secret created by Helm itself, the secret created on line 178 is for tracking the actual erro on install/upgrade . The same secret is also cleaned up in the getSecret utility.go method if error key is found in secret data.

Reply:
The way I understand it the secret is only created if commend returns an error. Normally the cmd.Run call will create the secret. This is an edge case.

Reply:
Okay, thanks for noting that this is created in the case an error occur!